src/test/java/songbox/house/threadchange/ThreadChangeTest.java

```diff
@@ -0,0 +1,148 @@
+package songbox.house.threadchange;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import songbox.house.util.ExecutorUtil;
+import songbox.house.util.ThreadChange;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.junit.Assert.*;
+
+public class ThreadChangeTest {
+    ExecutorService executorService = null;
+    static ThreadLocal<String> threadLocal = new ThreadLocal<>();
+    private class ThreadLocalThreadChangeListener implements ThreadChange.ThreadChangeListener {
+        Map<String, String> changeUUIDToThreadLocalValue = new ConcurrentHashMap<>();
+        @Override
+        public void didChange(String changeUUID) {
+            String localValue = changeUUIDToThreadLocalValue.get(changeUUID);
+            assertNotNull(localValue);
+
+            assertNull("Context is not cleared", threadLocal.get());
+            threadLocal.set(localValue);
+        }
+
+        @Override
+        public void willChange(String changeUUID) {
+            assertNotNull(threadLocal.get());
+            changeUUIDToThreadLocalValue.put(changeUUID, threadLocal.get());
+        }
+
+        @Override
+        public void finish(String changeUUID) {
+            assertNotNull(threadLocal.get());
+            threadLocal.set(null);
+            assertTrue(changeUUIDToThreadLocalValue.containsKey(changeUUID));
+            changeUUIDToThreadLocalValue.remove(changeUUID);
+        }
+    }
+
+    @Before
+    public void init() {
+        threadLocal.set(null);
+        executorService = ExecutorUtil.createExecutorService(4);
+    }
+
+    @Test
+    public void testThreadIsChangedAsRunnable() {
+        ThreadLocalThreadChangeListener threadLocalThreadChangeListener = new ThreadLocalThreadChangeListener();
+        ThreadChange.addThreadChangeListener(threadLocalThreadChangeListener);
+
+        String userName = UUID.randomUUID().toString();
+        threadLocal.set(userName);
+        System.out.println("Hello my name is: " + userName);
+
+        final boolean[] taskIsFinished = {false};
+        executorService.submit(new Runnable() {
+            @Override
+            public void run() {
+                assertEquals(userName, threadLocal.get());
+                System.out.println("Hello from another thread, my name is :" + threadLocal.get());
+                taskIsFinished[0] = true;
+            }
+        });
+
+
+        while (!taskIsFinished[0]) {
+            try {
+                Thread.sleep(100);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+        }
+
+        ThreadChange.removeThreadChangeListener(threadLocalThreadChangeListener);
+    }
+
+    @Test
+    public void testThreadIsChangedAsTask() {
```

Comment:
shouldPassThreadLocalParamsToNewCallable

Reply:
for me `test` reads better than methods with should, why it should pass test?

Reply:
For me it sounds like a tautology - you already have a class name with Test and annotation @Test. `should` because every test SHOULD pass, if it did not pass,- tested code is unstable and can't be used